### DISCUSSION TITLE
SQL Server Column types and nullability for the DataSync columns

### DISCUSSION THREAD

**Discussion by nicolasHul** [Azure/azure-mobile-apps#837] [Q&A] [answered]
Firstly, I would like to express my gratitude to @adrianhall for the exceptional work and assistance provided, including on Reddit.

I am currently attempting to implement DataSync on my server side and have encountered an issue when manually testing (via HTTP GET call) my sync endpoint. After some investigation, the issue may be related to the configuration of my entity.

For context, I have already set up my database tables in SQL Server. I manually added the following columns to the table:

- MobileId (Mapped to Id in my EF Core model. I used ‘MobileId’ instead of ‘Id’ as an ‘Id’ column already exists as an int column)
- UpdatedAt (datetimeoffset(7) NULL)
- Version (timestamp, NOT NULL)
- Deleted (bit NULL)

I am uncertain if I have correctly mapped the Data Types and nullability for these columns as I could not find any related information. I used the C# properties as a reference. 

I suspect that the `UpdatedAt` column should also be ‘NOT NULL’. When I inherit the `EntityTableData` class to my own entity class, I encounter EF Core NULL data errors.

I have the following questions:

- Have I overlooked any information or SQL scripts related to the table setup in SQL Server?
- Is it required for the `UpdatedAt` column to be NOT NULL? Should I populate this column for all existing records with the current UTCDATETIME (i.e., ‘SYSUTCDATETIME()’) in SQL Server?
- If `UpdatedAt` is required, does this imply that I need to set the default value for the column to ‘SYSUTCDATETIME()’ for backward compatibility with my other applications?

**Comment by adrianhall** [comment, marked as answer]
It's worth understanding how these fields are used in the application.

When a synchronization occurs, the client will send the following rough query:

```text
/tables/myTable?$filter=(updatedAt gt '&lt;last-date-of-sync&gt;')
```

The updatedAt field MUST be present when doing incremental synchronization.  NOT NULL is an excellent method for ensuring this, but I would set it to the epoch (Midnight on Jan-1-1970 UTC) as this would allow all records to be synchronized.  Alternatively, setting it to SYSUTCDATETIME() would only synchronize the records if you had never synchronized data before.

Version is a byte[] array and opaque.  We do a sequence comparison to ensure that we are not inadvertently changing data when we don't expect.  The term for this is "optimistic concurrency".  In EF Core, we use the Version column as the concurrency token to provide an extra level of protection.  In addition, we do ETag (If-Match and If-None-Match) processing on PUT and DELETE to ensure we don't overwrite data.

Deleted is a flag that allows us to inform other clients that the record has been deleted.  When we do the synchronization, we add a query parameter __includeDeleted=true to pull down records with the Deleted flag set; otherwise we do a search `AND Deleted = FALSE` on the end of the search.  This should be `Deleted (bit default 0)` and update all records to false when you update the database.

Now - why is DateTimeOffset nullable in the c# interface then?  Well, it's a value type and via a quirk in the JSON deserializer, it will cause problems when you submit a POST (to create an entity) to the table without it.  Thus we had to make it nullable in order to make the entity creation in a remote table work resiliently.

**Reply by adrianhall** [reply]
... and the short version is:

* Nope
* Yes, UpdatedAt should be NOT NULL; set to the epoch
* Yes, set UpdatedAt with a default value of the epoch.

**Reply by nicolasHul** [reply]
Thank you! This makes it entirely clear! After the modification(s) it is working properly.

**Reply by adrianhall** [reply]
In case it wasn't completely obvious, this ALSO means:

* UpdatedAt should be indexed
* Deleted should be indexed
* Timestamp does not HAVE to be indexed (but it doesn't hurt)

And, in your case, MobileId should be a primary unique key.
